pkg/engine/task/task_toggle_test.go

```diff
@@ -0,0 +1,248 @@
+package task
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	"github.com/kudobuilder/kudo/pkg/engine"
+	"github.com/kudobuilder/kudo/pkg/engine/renderer"
+	"github.com/kudobuilder/kudo/pkg/test/utils"
+)
+
+func TestToggleTask_Run(t *testing.T) {
+	meta := renderer.Metadata{
+		Metadata: engine.Metadata{
+			InstanceName:        "test",
+			InstanceNamespace:   "default",
+			OperatorName:        "first-operator",
+			OperatorVersionName: "first-operator-1.0",
+			OperatorVersion:     "1.0",
+		},
+		PlanName:  "plan",
+		PhaseName: "phase",
+		StepName:  "step",
+		TaskName:  "task",
+	}
+
+	tests := []struct {
+		name    string
+		task    ToggleTask
+		done    bool
+		wantErr bool
+		fatal   bool
+		ctx     Context
+	}{
+		{
+			name: "fails when no parameter is passed",
+			task: ToggleTask{
+				Name:      "task",
+				Resources: []string{"pod"},
+				Parameter: "feature-enabled",
+			},
+			done:    false,
+			wantErr: true,
+			fatal:   true,
+			ctx: Context{
+				Client:    fake.NewFakeClientWithScheme(scheme.Scheme),
+				Discovery: utils.FakeDiscoveryClient(),
+				Enhancer:  &testEnhancer{},
+				Meta:      meta,
+				Templates: map[string]string{},
+			},
+		},
+		{
+			name: "succeeds when the resource is healthy",
+			task: ToggleTask{
+				Name:      "task",
+				Resources: []string{"pod"},
+				Parameter: "feature-enabled",
+			},
+			done:    true,
+			wantErr: false,
+			ctx: Context{
+				Parameters: map[string]string{
+					"feature-enabled": "true",
+				},
+				Client:    fake.NewFakeClientWithScheme(scheme.Scheme),
+				Discovery: utils.FakeDiscoveryClient(),
+				Enhancer:  &testEnhancer{},
+				Meta:      meta,
+				Templates: map[string]string{"pod": resourceAsString(pod("pod1", "default"))},
+			},
+		},
+		{
+			name: "is not done when the resource is unhealthy",
```

Comment:
IMO we should be testing what is unique about toggle...  This test in particular is sketchy to me.    you have to go deep to really understand that you are using a "job" instead of a "pod" and then you have to understand that a job needs to be completed to be "healthy" none of which is important to testing toggle.   If we are to keep this (which I prefer to have it removed) it needs a lot of documentation to enable those that come across it to be successful with understanding the nuances.

IMO we should remove anything that doesn't test a code path through toggle.   We should have other tests to check health... and we should have e2e tests with toggle that tests deep if desired.

Reply:
there are two tests, one is `TestToggleTask_Run` which I think should be present because we have to test the `Run` which should make sure that we don't introduce any bugs in the `Run` method.
And the other test is `TestToggleTask_intermediateTask` which tests the whole logic of which task we should run, `Apply` or `Delete` 
There is nothing that doesn't go through Toggle path here, I just want to make sure that it is working as expected in the case of Job which is used in `Apply` tests also and I found it really a useful test to see the `Toggle` task status isn't done when resource is unhealthy.

Reply:
We've been using the `NewFakeClientWithScheme` and fake Jobs and Pods when testing other tasks too. Toggle task reuses Apply and Delete tasks internally but we still need to test the `Run` method 🤷‍♂